controllers/provider-vsphere/pkg/apis/vsphere/v1alpha1/types_cloudprofile.go

```diff
@@ -0,0 +1,179 @@
+	Folder string `json:"folder"`
+	// Regions is the specification of regions and zones topology
+	Regions []RegionSpec `json:"regions"`
+	// DefaultClassStoragePolicyName is the name of the vSphere storage policy to use for the 'default-class' storage class
+	DefaultClassStoragePolicyName string `json:"defaultClassStoragePolicyName"`
+	// FailureDomainLabels are the tag categories used for regions and zones.
+	// +optional
+	FailureDomainLabels *FailureDomainLabels `json:"failureDomainLabels,omitempty"`
+	// DNSServers is a list of IPs of DNS servers used while creating subnets.
+	DNSServers []string `json:"dnsServers"`
+	// MachineImages is the list of machine images that are understood by the controller. It maps
+	// logical names and versions to provider-specific identifiers.
+	MachineImages []MachineImages `json:"machineImages"`
+	// Constraints is an object containing constraints for certain values in the control plane config.
+	Constraints Constraints `json:"constraints"`
+}
+
+// FailureDomainLabels are the tag categories used for regions and zones in vSphere CSI driver and cloud controller.
+// See Cloud Native Storage: Set Up Zones in the vSphere CNS Environment
+// (https://docs.vmware.com/en/VMware-vSphere/6.7/Cloud-Native-Storage/GUID-9BD8CD12-CB24-4DF4-B4F0-A862D0C82C3B.html)
+type FailureDomainLabels struct {
+	// Region is the tag category used for region on vSphere data centers and/or clusters.
+	Region string `json:"region"`
+	// Zone is the tag category used for zones on vSphere data centers and/or clusters.
+	Zone string `json:"zone"`
+}
+
+// RegionSpec specifies the topology of a region and its zones.
+// A region consists of a Vcenter host, transport zone and optionally a data center.
+// A zone in a region consists of a data center (if not specified in the region), a computer cluster,
+// and optionally a resource zone or host system.
+type RegionSpec struct {
+	// Name is the name of the region
+	Name string `json:"name"`
+	// VsphereHost is the vSphere host
+	VsphereHost string `json:"vsphereHost"`
+	// VsphereInsecureSSL is a flag if insecure HTTPS is allowed for VsphereHost
+	VsphereInsecureSSL bool `json:"vsphereInsecureSSL"`
+	// NSXTHost is the NSX-T host
+	NSXTHost string `json:"nsxtHost"`
+	// NSXTInsecureSSL is a flag if insecure HTTPS is allowed for NSXTHost
+	NSXTInsecureSSL bool `json:"nsxtInsecureSSL"`
+	// TransportZone is the NSX-T transport zone
+	TransportZone string `json:"transportZone"`
+	// LogicalTier0Router is the NSX-T logical tier 0 router
+	LogicalTier0Router string `json:"logicalTier0Router"`
+	// EdgeCluster is the NSX-T edge cluster
+	EdgeCluster string `json:"edgeCluster"`
+	// SNATIPPool is the NSX-T IP pool to allocate the SNAT ip address
+	SNATIPPool string `json:"snatIPPool"`
+
+	// Datacenter is the name of the vSphere data center (data center can either be defined at region or zone level)
+	// +optional
+	Datacenter string `json:"datacenter,omitempty"`
```

Comment:
See above: Ensure all optional fields have pointer types (please double-check all your resources, not only this file).